app/models/user.rb

```diff
@@ -126,18 +116,36 @@ def hidden_posts
     blocked_or_hidden_posts('hidden', posts_blocked, full_blocked)
   end
 
-  private
+  # override devise to migrate passwords from legacy format
+  # https://github.com/heartcombo/devise/wiki/How-To:-Migration-legacy-database
+  def valid_password?(password)
+    # 2024+ password format
+    return super unless self.legacy_password_hash.present?
+
+    valid = if salt_uuid.present? # 2016-2024 password format
+      legacy_password_hash == crypted_password(password)
+    else # pre-2016 password format
+      legacy_password_hash == old_crypted_password(password)
+    end
+
+    if valid # migrate to new password format (devise, 2024+)
+      self.password = password
+      self.legacy_password_hash = nil
+      self.salt_uuid = nil
+      self.save!
+    end
```

Comment:
I think you could do this with a single `update!` call?

Reply:
True!

```suggestion
    if valid # migrate to new password format (devise, 2024+)
      self.update!(password: password, legacy_password_hash: nil, salt_uuid: nil)
    end
```

Reply:
I mean at that point you probably want
```suggestion
    # migrate to new password format (devise, 2025+)
    self.update!(password: password, legacy_password_hash: nil, salt_uuid: nil) if valid
```